Node Quorum

laumair opened the pull request:
# Description

Fixes # https://github.com/iotaledger/trinity-wallet/issues/453

Depends on https://github.com/iotaledger/iota.js/pull/314

- `wereAddressesSpentFrom` - Compute a quorum result from multiple nodes. Fallback to `true` if there is no consensus among results. 
- `getInclusionStates ` - First compute a quorum result for `latestSolidSubtangleMilestone`, then compute a quorum for inclusion states by using the agreed `latestSolidSubtangleMilestone` as a tip. Fallback to `false` if there is no consensus among results.
- `getBalances` - Similar to `getInclusionStates`. First compute a quorum result for `latestSolidSubtangleMilestone`, then compute a quorum for balances by using the agreed `latestSolidSubtangleMilestone` as a tip. Fallback to `'0'` if there is no consensus among results.
- ~~`getTrytes` - Compute a quorum result from multiple nodes. Fallback to null hash trytes if there is no consensus among results.~~ 
- ~~`findTransactions` - The strategy for determining a quorum result for this endpoint is different than the rest. Computing a quorum for`findTransactions` simply prepare a set of unique hashes from multiple nodes.~~
## Type of change

- Enhancement

# How Has This Been Tested?

**TODO**

# Checklist:

_Please delete items that are not relevant._

- [x] My code follows the style guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] For changes to `mobile` that include native code (including React Native modules): I have verified that both iOS and Android successfully build in both `Debug` and `Release` modes
- [ ] For changes to `shared`: If applicable, I have verified that my changes are implemented correctly in `desktop` and `mobile`

Commit:
Implement quorum for wereAddressesSpentFrom

Commit:
Simply #findSyncedNodes implementation

Commit:
Merge branch 'develop' into feat/quorum

Commit:
Merge branch 'develop' into feat/quorum

Commit:
Merge branch develop into branch feat/quorum

Commit:
Add quorum support for getBalances IRI endpoint

Commit:
Minor updates

- Add quorum support for getTrytes IRI endpoint
- Minor clean up in quorum methods

Commit:
Add quorum support for findTransactions IRI endpoint

Commit:
Rename quorum methods for better readability

Commit:
Integrate quorum methods with extended api

Commit:
Merge branch develop into branch feat/quorum

Commit:
Remove findTransactions & getTrytes endpoints from quorum

alon-e commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,415 @@
+ *
+ *   @method determineQuorumResult
+ *   @param {array} validResults
+ *   @param {number} faultyResultCount
```
I think this changed in the impl (to quorumSize), but not the JSDocs.

laumair replied:
Good catch. [Fixed](https://github.com/laumair/trinity-wallet/blob/feat/quorum/src/shared/libs/iota/quorum.js#L31).

alon-e commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,415 @@
+
+    return (method, threshold) => {
+        // Include faulty result count while determining percentage
+        const percentage = frequency[mostFrequent] / quorumSize * 100;
```
even though it evaluates correctly. I think this is clearer:
```suggestion
        const percentage = (frequency[mostFrequent] / quorumSize) * 100;
```

alon-e commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,415 @@
+ *
+ *   @returns {function}
+ **/
+const determineQuorumResult = (validResults, quorumSize = QUORUM_SIZE) => {
```
how are valid results defined? (what is an invalid result, if someone did not answer?)

laumair replied:
That's correct. Invalid results are basically the results from unresponsive nodes. 

If there's a network error, we explicitly mark the result as `undefined` [here](https://github.com/laumair/trinity-wallet/blob/feat/quorum/src/shared/libs/iota/quorum.js#L273) and later on filter these results.

alon-e replied:
To clarify that this is intentional (seems good to me)
the quorum is calculated *only* within the responsive nodes.
this should be mentioned somewhere - b/c it changes the strategy to attack the quorum.

for example:
```
N = 7
m = 3 (malicious)
scenario A:
all m vote No
all honest vote Yes
Result: 4/7= FailSafe

scenario B:
2 m vote "unresponsive"
1 m votes No
all honest vote Yes
Result: 3/4= Yes
```

so the best stratagy is to participate.
OTOH,
if the malicious nodes manage to kick off honest nodes (the will become uresponsive), they can pass bad elections:
```

scenario C:
3 m vote No
2 honest vote Yes
2 honest vote "unresponsive" due to malice
Result: 3/5= No
(well that's cutting it close, but you get the drift)
```

to give a full picture - anouther option would be to have "unresponsive" vote FailSafe.
but propbably needs some analysis as well.

laumair replied:
> the quorum is calculated only within the responsive nodes.

The quorum is not calculated within the responsive nodes. Unresponsive nodes would also be considered. Explicitly passing `undefined` as a result for unresponsive nodes is just to easily filter out invalid results. 

Quorum [calculation](https://github.com/laumair/trinity-wallet/blob/feat/quorum/src/shared/libs/iota/quorum.js#L73) would always be out of `N` (`N === quorumSize`). 

> scenario B:
2 m vote "unresponsive"
1 m votes No
all honest vote Yes
Result: 3/4= Yes

In this case, result would be 3/N. 

> anouther option would be to have "unresponsive" vote FailSafe.

IIUC, then always calculating quorum out of `N` basically does the same?

alon-e replied:
>The quorum is not calculated within the responsive nodes. Unresponsive nodes would also be considered. Explicitly passing undefined as a result for unresponsive nodes is just to easily filter out invalid results.

oh, so that is confusing -

> That's correct. Invalid results are basically the results from unresponsive nodes.

 the param `validResults` sounds like it's only the.. well ... valid results. filtering the unresponsive nodes.

alon-e replied:
oh, so `quorumSize = QUORUM_SIZE` is `N`, regardless of the size of `validResults` - :+1: 
maybe worth another pass to make it a bit clearer :)

laumair replied:
> oh, so quorumSize = QUORUM_SIZE is N, regardless of the size of validResults

That's correct. I'll make another commit to make it clearer. Sorry about the confusion.

laumair replied:
[This](https://github.com/iotaledger/trinity-wallet/pull/631/commits/d0a4dd9c0a12e24610e6238b13de71fce19a8543) should improve the readability of the overall implementation.

alon-e commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,415 @@
+    const whitelistedNodes = filter(nodes, (node) => !includes(blacklistedNodes, node) && !includes(syncedNodes, node));
+
+    if (isEmpty(whitelistedNodes)) {
+        return Promise.reject(new Error(Errors.NOT_ENOUGH_HEALTHY_QUORUM_NODES));
```
could it be that case that `syncedNodes` is large enough -> `|syncedNodes| = quorumSize`, and this is the reason that `isEmpty(whitelistedNodes)` ?

laumair replied:
That was actually a minor bug. Thank you. 

The only reason why this should reject is if there no whitelisted nodes and we still need to fill up _synced_ nodes to reach the quorum size. Have fixed it [here](https://github.com/laumair/trinity-wallet/blob/feat/quorum/src/shared/libs/iota/quorum.js#L127) .

laumair replied:
Also added [test coverage](https://github.com/laumair/trinity-wallet/blob/feat/quorum/src/shared/__tests__/libs/iota/quorum.spec.js#L144) for this.

alon-e commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,415 @@
+ *
+ *   @returns {Promise}
+ **/
+const getQuorumForWereAddressesSpentFrom = (payload, syncedNodes) => {
```
the following 4 (?) functions are duplicates except for the method name (which is used to invoke the API and get the safe default). 

could this be done using the `method` as an argument? and have only 1 copy of this function, the reason being that this mechanism is new and might need to be tweaked, maintaining all the copies is a hassle and would lead to errors.


and later adding more methods to quorum would be easier as well.

laumair replied:
Great suggestion. 

One reason why I avoided doing that in the first place is because some methods (e.g. getBalances, getNodeInfo) response is not a plain array of booleans like the rest of the methods (wereAddressesSpentFrom, getLatestInclusion). However, it's not a big problem. I agree that all four methods had duplication and maintaining all four would be a hassle. 

Have simplified the implementation. Now, there's only a single [function](https://github.com/laumair/trinity-wallet/blob/feat/quorum/src/shared/libs/iota/quorum.js#L261) that serves all four endpoints and also a separate [function](https://github.com/laumair/trinity-wallet/blob/feat/quorum/src/shared/libs/iota/quorum.js#L180) for preparing results. 

Reason why I decoupled preparation of results from `#getQuorum` is because it would be a lot easier to unit test both separately.

alon-e commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,415 @@
+
+            while (idx < requestPayloadSize) {
+                /* eslint-disable no-loop-func */
+                quorumResult.push(
```
IIUC, if the request is for L addresses the Quorom is determined for each address separately, correct?
if so, that's great! (what I envisioned) but it should be clearly stated (in the JSDocs?)

laumair replied:
That's correct. Quorum is determined for each address separately. 
Have added better code comments and also have updated [JSDocs](https://github.com/laumair/trinity-wallet/blob/feat/quorum/src/shared/libs/iota/quorum.js#L167-L171).

alon-e commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,415 @@
+                    syncedNodes,
+                    (provider) =>
+                        new Promise((resolve) => {
+                            new IOTA({ provider }).api.wereAddressesSpentFrom(
```
does this have some type of timeout? so err (or valid response) is promised to be returned eventually?

laumair replied:
It will eventually timeout (because of the addition [here)](https://github.com/laumair/trinity-wallet/blob/feat/quorum/src/shared/libs/iota/quorum.js#L272) and an error or a valid response is guaranteed.

alon-e commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,415 @@
+         */
+        getLatestInclusion(hashes) {
+            return findSyncedNodesIfNecessary().then((newSyncedNodes) =>
+                getQuorumForLatestSolidSubtangleMilestone(newSyncedNodes).then((latestSolidSubtangleMilestone) =>
```
this addition is worth mentioning somewhere - in the JSDocs (same for `getBalances`)

laumair replied:
Did you mean documentation for first getting quorum results for `latestSolidSubtangleMilestone`? Have updated [JSDocs](https://github.com/laumair/trinity-wallet/blob/feat/quorum/src/shared/libs/iota/quorum.js#L345-L347) for this method also for [getBalances](https://github.com/laumair/trinity-wallet/blob/feat/quorum/src/shared/libs/iota/quorum.js#L368-L370).

alon-e commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,415 @@
+            const validResults = filter(results, (result) => !isUndefined(result));
+
+            // Get quorum result for latest solid subtangle milestone
+            const latestSolidSubtangleMilestone = determineQuorumResult(
```
nit-pick, but this should follow the structure of the code above, it makes maintenance harder (I understand that there is no payload and the response is a single value)

laumair replied:
Good suggestion. As mentioned [here](https://github.com/iotaledger/trinity-wallet/pull/631#discussion_r241632229), the pattern is now consistent with these [changes](https://github.com/laumair/trinity-wallet/blob/feat/quorum/src/shared/libs/iota/quorum.js#L180-L284).

alon-e commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,415 @@
+            // If nodes cannot agree on the latestSolidSubtangleMilestone
+            // Then just throw an exception
+            if (latestSolidSubtangleMilestone === EMPTY_HASH_TRYTES) {
+                // TODO (laumair): Might need to replace the error message here.
```
good idea :)

alon-e reviewed (changes requested):
looks great!
requested changes mostly b/c comment doesn't affect review count.

my main open question is if there is a timeout for unresponsive nodes (to mark invalid).

Commit:
Refactor quorum implementation and do minor fixes

- Update JSDoc typos
- Simplify quorum implementation (Remove duplications)
- Add a timeout for network request to each node
- Fix issues in findSyncedNodes implementation
- Update error messages

Commit:
Add coverage
- Add coverage for #determineQuorumResult
- Add coverage for #fallbackToSafeResult
- Add coverage for #findSyncedNodes

Commit:
Add empty payload checks in quorum methods

laumair commented:
@alon-e Thank you for your review. 

> my main open question is if there is a timeout for unresponsive nodes (to mark invalid).

This was indeed missing. Have fixed it [here](https://github.com/laumair/trinity-wallet/blob/feat/quorum/src/shared/libs/iota/quorum.js#L272). [getIotaInstance](https://github.com/laumair/trinity-wallet/blob/feat/quorum/src/shared/libs/iota/extendedApi.js#L31) sets the default timeout.

alon-e committed:
Wrap percentage calculation in parentheses for clarity

Co-Authored-By: laumair <aquadestructor@icloud.com>

Commit:
Merge upstream branch develop into origin branch develop

Commit:
Merge branch 'feat/quorum' of github.com:laumair/trinity-wallet into feat/quorum

Commit:
Use develop branch of iota.lib.js

Commit:
Enforce quorum (by default) on supported methods

Commit:
Fix tests failing because of quorum enforcement

Commit:
Add code documentation and rename parameters & variables for clarity

- Related discussion https://github.com/iotaledger/trinity-wallet/pull/631#discussion_r241307209

dependabot[bot] committed:
[Security] Bump cryptiles from 3.1.2 to 3.1.4 (#829)

dependabot[bot] committed:
[Security] Bump nokogiri from 1.8.4 to 1.9.1 in /src/mobile/android (#828)

Commit:
Make sure accounts are always iterated in correct order (by account index) (#824)

Object.keys(<object>) function does not always preseve the order, especially if the object key starts with a number. This causes an issue when Object.keys is used for iterating on account names. https://github.com/iotaledger/trinity-wallet/pull/715 adds account indexes to state to make sure the order of accounts is always intact. However, some components in desktop use Object.keys directly on accounts object, which leads to certain issues of incorrect references to accounts. This commit fixes this issue by replacing Object.keys implementation on accounts with getAccountNamesFromState selector that guarantees the accounts order.

Fixes https://github.com/iotaledger/trinity-wallet/issues/811

Note that the issues Object.keys create are not always noticeable. Steps to reproduce these issues are:

- Add account with name "M"
- Add another account with name "0"
- Notice account names order in sidebar (Instead of "0" being the second account, it becomes the first)
- Generate receive address from account "M" (Instead of generating receive address for account "M", it generates receive address for account "0")

iotasyncbot committed:
New Crowdin translations [ci skip] (#826)

* New translations translation.json (Czech)

* New translations translation.json (Czech)

* New translations translation.json (Japanese)

* New translations translation.json (Spanish)

cvarley100 commented on src/shared/config.js:
```diff
@@ -44,3 +44,7 @@ export const NODE_REQUEST_TIMEOUT = 6000 * 2;
 export const IRI_API_VERSION = '1';
+
+export const QUORUM_THRESHOLD = 67;
+export const QUORUM_NODES_SIZE = 7;
```
I think we can change the variable names here
```suggestion
export const QUORUM_SIZE = 7;
```

laumair replied:
Resolved [here](https://github.com/iotaledger/trinity-wallet/pull/631/commits/d8ad41e6a738ac82b528e3fc7ef87ee6225b956b#diff-78157a2166aa6eaa7e812e543c8729ddR15)

cvarley100 commented on src/shared/config.js:
```diff
@@ -44,3 +44,7 @@ export const NODE_REQUEST_TIMEOUT = 6000 * 2;
+
+export const QUORUM_THRESHOLD = 67;
+export const QUORUM_NODES_SIZE = 7;
+export const QUORUM_NODES_SYNC_CHECKS_INTERVAL = 120;
```
```suggestion
export const QUORUM_SYNC_CHECK_INTERVAL = 120;
```

laumair replied:
Resolved [here](https://github.com/iotaledger/trinity-wallet/pull/631/commits/d8ad41e6a738ac82b528e3fc7ef87ee6225b956b#diff-78157a2166aa6eaa7e812e543c8729ddR15)

cvarley100 commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,407 @@
+import {
+    QUORUM_THRESHOLD,
+    QUORUM_NODES_SIZE,
+    QUORUM_NODES_SYNC_CHECKS_INTERVAL,
```
```suggestion
    QUORUM_SYNC_CHECK_INTERVAL,
```

laumair replied:
Resolved [here](https://github.com/iotaledger/trinity-wallet/pull/631/commits/d8ad41e6a738ac82b528e3fc7ef87ee6225b956b#diff-78157a2166aa6eaa7e812e543c8729ddR15)

cvarley100 commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,407 @@
+import { isNodeHealthy, getIotaInstance } from './extendedApi';
+import {
+    QUORUM_THRESHOLD,
+    QUORUM_NODES_SIZE,
```
```suggestion
    QUORUM_SIZE,
```

cvarley100 commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,407 @@
+ * @param {number} quorumSize
+ * @returns {Promise}
+ */
+const rejectIfNotEnoughSyncedNodes = (nodes, quorumSize = QUORUM_NODES_SIZE) => {
```
```suggestion
const rejectIfNotEnoughSyncedNodes = (nodes, quorumSize = QUORUM_SIZE) => {
```

laumair replied:
Resolved [here](https://github.com/iotaledger/trinity-wallet/pull/631/commits/d8ad41e6a738ac82b528e3fc7ef87ee6225b956b#diff-78157a2166aa6eaa7e812e543c8729ddR29)

cvarley100 commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,407 @@
+    const findSyncedNodesIfNecessary = () => {
+        const timeElapsed = (new Date() - lastSyncedAt) / 1000;
+
+        if (isEmpty(syncedNodes) || timeElapsed >= QUORUM_NODES_SYNC_CHECKS_INTERVAL) {
```
```suggestion
        if (isEmpty(selectedNodes) || timeElapsed >= QUORUM_SYNC_CHECK_INTERVAL) {
```

laumair replied:
Resolved [here](https://github.com/iotaledger/trinity-wallet/pull/631/commits/d8ad41e6a738ac82b528e3fc7ef87ee6225b956b#diff-78157a2166aa6eaa7e812e543c8729ddR310)

cvarley100 commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,407 @@
+ *
+ * @returns {Promise<string>}
+ */
+const rejectIfEmptyHashTrytes = (latestSolidSubtangleMilestone) => {
```
This is a weird method name

laumair replied:
It rejects a promise if provided trytes are null hash trytes. Have changed the [paramter](https://github.com/iotaledger/trinity-wallet/pull/631/commits/d8ad41e6a738ac82b528e3fc7ef87ee6225b956b#diff-78157a2166aa6eaa7e812e543c8729ddR46) to just `trytes` for clarity

cvarley100 commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,407 @@
+ *
+ *   @returns {function(string, [number]): {boolean | string}}
+ **/
+const determineQuorumResult = (validResults, quorumSize = QUORUM_NODES_SIZE) => {
```
```suggestion
const determineQuorumResult = (validResults, quorumSize = QUORUM_SIZE) => {
```

cvarley100 commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,407 @@
+};
+
+/**
+ *   For a list of nodes, find healthy (synced) nodes N, where size(N) === QUORUM_NODES_SIZE
```
```suggestion
 *   For a list of nodes, find healthy (synced) nodes N, where size(N) === QUORUM_SIZE
```

cvarley100 commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,407 @@
+export default function Quorum(quorumNodes) {
+    let nodes = uniq(quorumNodes);
+
+    if (size(nodes) < QUORUM_NODES_SIZE) {
```
```suggestion
    if (size(nodes) < QUORUM_SIZE) {
```

cvarley100 commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,407 @@
+        const timeElapsed = (new Date() - lastSyncedAt) / 1000;
+
+        if (isEmpty(syncedNodes) || timeElapsed >= QUORUM_NODES_SYNC_CHECKS_INTERVAL) {
+            return findSyncedNodes(nodes, QUORUM_NODES_SIZE, syncedNodes).then((newSyncedNodes) => {
```
```suggestion
            return findSyncedNodes(nodes, QUORUM_SIZE, selectedNodes).then((syncedNodes) => {
```

cvarley100 commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,407 @@
+
+    return (method, threshold = QUORUM_THRESHOLD) => {
+        // Always calculate percentage out of quorum size i.e., all (responsive + unresponsive) nodes.
+        const percentage = frequency[mostFrequent] / quorumSize * 100;
```
We could remove `100` and just do quorum threshold as a decimal

cvarley100 commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,407 @@
+    };
+
+    if (!includes(keys(allowedMethodsMap), method)) {
+        throw new Error(Errors.METHOD_NOT_SUPPORTED_FOR_QUORUM);
```
👍

cvarley100 committed:
Mobile: Update findSyncedNodes

cvarley100 committed:
Update quorum.js

cvarley100 commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,407 @@
+        throw new Error(Errors.NOT_ENOUGH_QUORUM_NODES);
+    }
+
+    let syncedNodes = [];
```
```suggestion
    let selectedNodes = [];
```

cvarley100 commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,407 @@
+
+        if (isEmpty(syncedNodes) || timeElapsed >= QUORUM_NODES_SYNC_CHECKS_INTERVAL) {
+            return findSyncedNodes(nodes, QUORUM_NODES_SIZE, syncedNodes).then((newSyncedNodes) => {
+                syncedNodes = newSyncedNodes;
```
```suggestion
                selectedNodes = syncedNodes;
```

cvarley100 commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,407 @@
+                syncedNodes = newSyncedNodes;
+                lastSyncedAt = new Date();
+
+                return newSyncedNodes;
```
```suggestion
                return syncedNodes;
```

cvarley100 commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,407 @@
+            });
+        }
+
+        return Promise.resolve(syncedNodes);
```
```suggestion
        return Promise.resolve(selectedNodes);
```

cvarley100 commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,407 @@
+        wereAddressesSpentFrom(addresses) {
+            return isEmpty(addresses)
+                ? Promise.resolve([])
+                : findSyncedNodesIfNecessary().then((newSyncedNodes) =>
```
```suggestion
                : findSyncedNodesIfNecessary().then((syncedNodes) =>
```

cvarley100 commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,407 @@
+            return isEmpty(addresses)
+                ? Promise.resolve([])
+                : findSyncedNodesIfNecessary().then((newSyncedNodes) =>
+                      getQuorum('wereAddressesSpentFrom', newSyncedNodes, addresses),
```
```suggestion
                      getQuorum('wereAddressesSpentFrom', syncedNodes, addresses),
```

cvarley100 commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,407 @@
+        getLatestInclusion(hashes) {
+            return isEmpty(hashes)
+                ? Promise.resolve([])
+                : findSyncedNodesIfNecessary().then((newSyncedNodes) =>
```
```suggestion
                : findSyncedNodesIfNecessary().then((syncedNodes) =>
```

cvarley100 commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,407 @@
+                          // No need to proceed further.
+                          .then(rejectIfEmptyHashTrytes)
+                          .then((latestSolidSubtangleMilestone) =>
+                              getQuorum('getInclusionStates', newSyncedNodes, hashes, [latestSolidSubtangleMilestone]),
```
```suggestion
                              getQuorum('getInclusionStates', syncedNodes, hashes, [latestSolidSubtangleMilestone]),
```

cvarley100 commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,407 @@
+            return isEmpty(addresses)
+                ? Promise.resolve([])
+                : findSyncedNodesIfNecessary().then((newSyncedNodes) =>
+                      getQuorum('getNodeInfo:latestSolidSubtangleMilestone', newSyncedNodes)
```
```suggestion
                      getQuorum('getNodeInfo:latestSolidSubtangleMilestone', syncedNodes)
```

laumair replied:
Resolved [here](https://github.com/laumair/trinity-wallet/blob/feat/quorum/src/shared/libs/iota/quorum.js#L372)

cvarley100 commented on src/shared/libs/iota/quorum.js:
```diff
@@ -0,0 +1,407 @@
+                          // No need to proceed further.
+                          .then(rejectIfEmptyHashTrytes)
+                          .then((latestSolidSubtangleMilestone) =>
+                              getQuorum('getBalances:balances', newSyncedNodes, addresses, threshold, [
```
```suggestion
                              getQuorum('getBalances:balances', syncedNodes, addresses, threshold, [
```

laumair replied:
Resolved [here](https://github.com/iotaledger/trinity-wallet/pull/631/commits/d8ad41e6a738ac82b528e3fc7ef87ee6225b956b#diff-78157a2166aa6eaa7e812e543c8729ddR396)

Commit:
Merge branch 'develop' into feat/quorum

Commit:
Merge branch 'feat/quorum' of github.com:laumair/trinity-wallet into feat/quorum

Commit:
Improve parameters and variable names

Commit:
Include custom nodes in quorum nodes

Commit:
Reduce node request timeout for getNodeInfo api calls (in quorum)
